### PR TITLE
sticky footer and projects_controller_spec

### PR DESCRIPTION
**Pull request by Kasvit**
**STATUS**
READY

**DESCRIPTION**
Added sticky footer and projects_controller_spec test + bug fix in projects controller

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [ ] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [ ] PR doesn't have conflicts with main branch
- [ ] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs

### CODE REVIEW

**Comment by Kasvit** [comment]
image
image
image

File: app/controllers/account/projects_controller.rb [hunk 1]
```diff
@@ -6,17 +6,19 @@ def index
 
   def show
     @workspace = parent
-    @project = @workspace.projects.find(params[:id])
+    @project = resource
```
Comment (BohdanBusko) [inline app/controllers/account/projects_controller.rb:4]: Почитай останні коменти до мого першого пулу так не добре робити

File: spec/controllers/account/projects_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  let!(:project_valid_params) { { name: "test_project"} }
+  let!(:project_invalid_params) { { name: nil} }
+
+  before do
+    sign_in user
+  end
+
+  describe "GET #index in Account/Projects controller" do
+    it "returns http success" do
+      get :index, params: { workspace_id: workspace.id }
```
Comment (BohdanBusko) [inline spec/controllers/account/projects_controller_spec.rb]: Перероби на workspace.to_param

File: app/assets/stylesheets/main.scss [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  padding-bottom: 50px;
+}
+.site-footer {
+  position: absolute;
+  bottom: 0px;
+  width: 100%;
+}
+.footer-content {
+  padding: 10px 10px;
+
```
Comment (igorkasyanchuk) [inline app/assets/stylesheets/main.scss]: why this new line is needed?

File: app/views/layouts/application.html.slim [hunk 4]
```diff
@@ -1,5 +1,5 @@
 doctype html
-html
+html.site
```
Comment (igorkasyanchuk) [inline app/views/layouts/application.html.slim]: не чіпляй красли до html

Comment (igorkasyanchuk) [reply app/views/layouts/application.html.slim]: до боді це найшвидше що можна

File: app/assets/stylesheets/main.scss (hunk 3 above)
Comment (Kasvit) [reply app/assets/stylesheets/main.scss]: image

File: spec/controllers/account/projects_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  let!(:workspace) { create(:workspace, user: user) }
+  let!(:project) { create(:project, workspace: workspace) }
+  let!(:project_valid_params) { { name: "test_project"} }
+  let!(:project_invalid_params) { { name: nil} }
+
+  before do
+    sign_in user
+  end
+
+  describe "GET #index in Account/Projects controller" do
```
Comment (DimaSmagulov) [inline spec/controllers/account/projects_controller_spec.rb]: we do not need `Account/Projects controller` in description those specs are already for it

File: spec/controllers/account/projects_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    it "returns http success" do
+      get :index, params: { workspace_id: workspace.to_param }
+      expect(response).to have_http_status(200)
+    end
+  end
+    describe "GET show in Account/Projects controller" do
```
Comment (DimaSmagulov) [inline spec/controllers/account/projects_controller_spec.rb]: why do #show in #index scope?

File: spec/controllers/account/projects_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  end
+
+  describe "DELETE #destroy in Account/Projects controller" do
+    it "returns http success" do
+      delete :destroy, params: { workspace_id: workspace.to_param, id: project.to_param }
+      expect(Project.find_by(id: project.to_param)).to eq(nil)
+    end
+  end
+
+end
```
Comment (DimaSmagulov) [inline spec/controllers/account/projects_controller_spec.rb:94]: Please add specs when user want to get access to projects from other account and when the user is not signed in.

**Review by DimaSmagulov — CHANGES_REQUESTED** [review]

File: app/controllers/account/projects_controller.rb (hunk 1 above)
Comment (romaslmd) [reply app/controllers/account/projects_controller.rb:4]: it's pretty ok to do like that, except the ruby instance method caching is needed, 
e.g.:
```
def parent
  @parent ||= SomeModel.find(params[:parent_id])
end

def resource
  @resource ||= parent.resources.find(params[:id])
end
```

File: app/views/layouts/application.html.slim (hunk 4 above)
Comment (romaslmd) [inline app/views/layouts/application.html.slim]: huh?

File: app/assets/stylesheets/main.scss [hunk 8]
```diff
@@ -0,0 +1,14 @@
+.site {
+  position: relative;
+  min-height: 100%;
+  padding-bottom: 50px;
```
Comment (romaslmd) [inline app/assets/stylesheets/main.scss]: why are u styling html?

File: app/controllers/account/projects_controller.rb [hunk 9]
```diff
@@ -6,45 +6,51 @@ def index
...
   end
 
   def update
+    @workspace = parent
     @project = resource
-
     if @project.update(project_params)
-      redirect_to account_workspace_projects_path(parent)
+      redirect_to account_workspace_projects_path(parent), notice: "Project was edited!"
```
Comment (romaslmd) [inline app/controllers/account/projects_controller.rb]: updated would sound more truly

File: spec/controllers/account/projects_controller_spec.rb (hunk 6 above)
Comment (romaslmd) [reply spec/controllers/account/projects_controller_spec.rb]: it's not, just wrong indentation

File: spec/controllers/account/projects_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    context "with walid attributes" do
+      it "must create new project" do
+        expect { post :create, params: { workspace_id: workspace.to_param, project: project_valid_params } }.to change(Project, :count).by(1)
+      end
+    end
+  
+    context "with inwalid attributes" do
+      it "must not create a project and render 'new' form" do
+        count_before = Project.count
+        post :create, params: { workspace_id: workspace.to_param, project: project_invalid_params }
```
Comment (romaslmd) [inline spec/controllers/account/projects_controller_spec.rb]: go w/ the same style,
`expect { post(:create, params: { workspace_id: workspace.to_param, project: project_invalid_params }) }.to change(Project, :count).by(0)`

File: spec/controllers/account/projects_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        expect(Project.count).to eq(count_before)
+        expect(response).to render_template("account/projects/_form")
```
Comment (romaslmd) [inline spec/controllers/account/projects_controller_spec.rb]: just render new is enough

File: spec/controllers/account/projects_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        put :update, params: { workspace_id: workspace.to_param, id: project.to_param, project: project_valid_params}
+        project.reload
+        expect(project.name).to eq(project_valid_params[:name])
+      end
+    end
+
+    context "with inwalid attributes" do
+      it "must not create update the project and render 'edit' form" do
+        put :update, params: { workspace_id: workspace.to_param, id: project.to_param, project: project_invalid_params}
+        expect(project.name).to eq(project.name)
```
Comment (romaslmd) [inline spec/controllers/account/projects_controller_spec.rb]: wuUut?

File: spec/controllers/account/projects_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        expect(response).to render_template(:edit)
+        expect(response).to render_template("account/projects/_form")
+      end
+    end
+  end
+
+  describe "DELETE #destroy in Account/Projects controller" do
+    it "returns http success" do
+      delete :destroy, params: { workspace_id: workspace.to_param, id: project.to_param }
```
Comment (romaslmd) [inline spec/controllers/account/projects_controller_spec.rb]: same style
```
expect { delete(:destroy, params: { workspace_id: workspace.to_param, id: project.to_param })  }.to change(Project, :count).by(-1)
```

File: spec/controllers/account/projects_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    context "with walid attributes" do
+      it "must update the project" do
+        put :update, params: { workspace_id: workspace.to_param, id: project.to_param, project: project_valid_params}
+        project.reload
+        expect(project.name).to eq(project_valid_params[:name])
+      end
+    end
+
+    context "with inwalid attributes" do
+      it "must not create update the project and render 'edit' form" do
```
Comment (romaslmd) [inline spec/controllers/account/projects_controller_spec.rb]: > must not create update the project

isn't that too complicated?

File: app/assets/stylesheets/main.scss (hunk 3 above)
Comment (igorkasyanchuk) [reply app/assets/stylesheets/main.scss]: I mean empty line in CSS

File: app/controllers/account/projects_controller.rb (hunk 1 above)
Comment (igorkasyanchuk) [reply app/controllers/account/projects_controller.rb:4]: actually caching is not needed. this is something what you calling one time. And don't introduce new variables.

Code is OK (almost)

parent is not a user, but Workspace

def parent
   current_user.workspace.find...
end

def resource
  parent.projects.find ....
end

File: app/controllers/account/projects_controller.rb (hunk 9 above)
Comment (igorkasyanchuk) [reply app/controllers/account/projects_controller.rb]: "Project was successfully updated"

Do similar to other places of App. Be consistent

File: spec/controllers/account/projects_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        expect(response).to render_template(:edit)
+        expect(response).to render_template("account/projects/_form")
+      end
+    end
+  end
+
+  describe "DELETE #destroy in Account/Projects controller" do
+    it "returns http success" do
+      delete :destroy, params: { workspace_id: workspace.to_param, id: project.to_param }
+      expect(Project.find_by(id: project.to_param)).to eq(nil)
```
Comment (igorkasyanchuk) [inline spec/controllers/account/projects_controller_spec.rb]: why project.to_param?

use just `.id`

File: spec/controllers/account/projects_controller_spec.rb (hunk 7 above)
Comment (igorkasyanchuk) [reply spec/controllers/account/projects_controller_spec.rb:94]: At least one spec for "index" action is ok for now. When user is not logged in. He must receive a 404 status code error, or redirect to login page.

**Review by igorkasyanchuk — CHANGES_REQUESTED** [review]

**Review by igorkasyanchuk — APPROVED** [review]

**Pull request merged**